### PR TITLE
document AbstractIrrational interface

### LINKED ISSUE #26550: Next steps/guidelines for AbstractIrrational
I am very excited about `AbstractIrrational` and am hoping to understand the intended guidelines for implementing subtypes of it.  The abstract type itself was introduced by @stevengj in #24245, e.g. for representing exact multiples of π.  @Jutho also mentioned a use case for square roots of rational numbers.

One concern/confusion I have is that in the both cases, not all possible values are actually irrational.  For instance, 0π is rational, as is √4.  Are types that derive from `AbstractIrrational` supposed to error when such values are given?

In my (tentative) opinion, it will be much more elegant for these types to allow *any* multiple of π, including 0π, as well as the square root of any rational number, including perfect squares.  However, this suggests that a value represented by `AbstractIrrational` need not actually be irrational, and thus some of the overloads introduced in #24245 are not true in general: for instance, it is claimed that `==(x::AbstractIrrational, y::Real) = false` because "Irrationals, by definition, can't have a finite representation equal them exactly."  Other comparison functions, as well as `iszero` and `isone`, have similar issues in general.

Assuming it is agreed that even rational values can be represented by an `AbstractIrrational`, one path forward is to narrow the definition of some of the above methods to be only for `Irrational` rather than `AbstractIrrational`.  And in any case, some documentation/guidelines on what methods somebody should overload for a new `AbstractIrrational` type would be very helpful.

### PR DESCRIPTION
**Pull request by stevengj**
Fixes #26550.

### CODE REVIEW

**Review by ararslan — APPROVED** [review]

**Comment by garrison** [comment]
Thanks for this.  I expect I will have some additional comments on this/#26550, but I am unable to type them up at the moment.  (Please wait before merging.)

**Comment by stevengj** [comment]
Bump @garrison

File: base/irrationals.jl [hunk 1]
```diff
@@ -5,7 +5,16 @@
...
+Number type representing an exact irrational value, which is automatically rounded to the correct precision in
+arithmetic operations with other numeric quantities.
+
+Subtypes `MyIrrational <: AbstractIrrational` should implement at least `==(::MyIrrational, ::MyIrrational)`,
+`hash(x::MyIrrational, h::UInt)`, and `convert(::Type{F}, x::MyIrrational) where {F <: Union{BigFloat,Float32,Float64}}`.
+
+If a subtype is used to represent values that may occasionally be rational (e.g. a square-root type that represents `√n`
+for integers `n` will give a rational result when `n` is a perfect square), then it should also implement
+`isinteger`, `iszero`, `isone`, and `==` with `Real` values (since all of these default to `false` for
+`AbstractIrrational` types).
```
Comment (garrison) [inline base/irrationals.jl]: It is probably worth mentioning here that `hash` for these types should be implemented such that the hash for rational values matches that given by other ways of representing the same value.

**Comment by kshyatt** [comment]
What's the status here?

**Comment by ViralBShah** [comment]
Can we merge this one?

**Comment by garrison** [comment]
I was originally very excited about the introduction of `AbstractIrrational` but later [questioned whether it is actually as useful as it first seemed](https://github.com/JuliaLang/julia/issues/26550#issuecomment-377611511).  But either way, it is in 1.0 and therefore should be documented regardless.  So I am in favor of this being merged.

**Review by garrison — APPROVED** [review]

**Pull request merged**